### PR TITLE
object: remove ObjectIO

### PR DESCRIPTION
**Pull request by lucasoshiro**
Removing ObjectIO and replacing it by Object o => IO o, as discussed in https://github.com/lucasoshiro/oshit/discussions/7.

### CODE REVIEW

File: src/Commands/Plumbing/Merge/MergeTree.hs [hunk 1]
```diff
@@ -9,9 +9,9 @@ import qualified Data.ByteString.Char8      as B
 cmdMergeTree :: Command
 cmdMergeTree (o:a:b:_) = do
   let (hashO, hashA, hashB) = (B.pack o, B.pack a, B.pack b)
-  let treeO = loadTree hashO
-  let treeA = loadTree hashA
-  let treeB = loadTree hashB
+  let treeO = loadObject hashO
+  let treeA = loadObject hashA
+  let treeB = loadObject hashB
```
Comment (d3adb5) [inline src/Commands/Plumbing/Merge/MergeTree.hs:14]: I feel these lines repeat themselves unnecessarily. Might be worth merging with the one above, too. What do you think?

```suggestion
  let (treeO:treeA:treeB:_) = map loadObject [hashO, hashA, hashB]
```

**Review by d3adb5 — APPROVED** [review]

**Comment by d3adb5** [comment]
I did just refactor `Core.Object` in #5. It should be trivial to rebase my changes off this once it's merged, though. _Should_ be, not sure if it will.

File: src/Commands/Plumbing/Merge/MergeTree.hs (hunk 1 above)
Comment (lucasoshiro) [reply src/Commands/Plumbing/Merge/MergeTree.hs:14]: That leads to this warning:

Captura de Tela 2023-03-08 às 16 05 22

I think ghc isn't smart enough to know that this list will always have only three elements...

File: src/Commands/Plumbing/Merge/MergeCommit.hs [hunk 2]
```diff
@@ -8,10 +8,10 @@ import Data.Time
 
 import qualified Data.ByteString.Char8      as B
 
-treeFromCommit :: Hash -> TreeIO
+treeFromCommit :: Hash -> IO Tree
 treeFromCommit hash = do
-  commit <- loadCommit hash
-  loadTree $ treeHash commit
+  commit <- loadObject hash :: IO Commit
```
[the remark this answers is not in the document]
Comment (lucasoshiro) [reply src/Commands/Plumbing/Merge/MergeCommit.hs]: Ok, done in b6c967685db738666ec31f9b8fd4fa8e5ca910ca

File: src/Commands/Plumbing/Merge/MergeCommit.hs [hunk 3]
```diff
@@ -8,10 +8,10 @@ import Data.Time
 
 import qualified Data.ByteString.Char8      as B
 
-treeFromCommit :: Hash -> TreeIO
+treeFromCommit :: Hash -> IO Tree
 treeFromCommit hash = do
```
[the remark this answers is not in the document]
Comment (lucasoshiro) [reply src/Commands/Plumbing/Merge/MergeCommit.hs]: Done in c2a82df5e9c404390635b287535f726a28ff80fa

File: src/Commands/Plumbing/Merge/MergeTree.hs (hunk 1 above)
Comment (d3adb5) [reply src/Commands/Plumbing/Merge/MergeTree.hs:14]: Fair enough. We can at least lower the number of lines in this `do` block as such:

```haskell
cmdMergeTree :: Command
cmdMergeTree (o:a:b:_) = do
  let (treeO, treeA, treeB) = (hashToTree o, hashToTree a, hashToTree b)
      hashToTree = loadTree . B.pack

  merged <- mergeTree treeO treeA treeB
  putStrLn $ B.unpack . hashObject $ merged
```

What do you think? I'll leave it as a suggestion, but approve the PR. :P

**Review by d3adb5 — APPROVED** [review]

File: src/Commands/Plumbing/Merge/MergeTree.hs (hunk 1 above)
Comment (lucasoshiro) [reply src/Commands/Plumbing/Merge/MergeTree.hs]: This works. Done in 93b5c6008210623970dbef81c0a33a2a690e450a.

Probably this code could be a lot cleaner using things like that.

**Comment by lucasoshiro** [comment]
Thanks @d3adb5 ! I'm merging it.

**Pull request merged**